crates/ruff/src/checkers/ast/analyze/expression.rs

```diff
@@ -1103,6 +1105,15 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
             {
                 flake8_pyi::rules::redundant_literal_union(checker, expr);
             }
+            if checker.enabled(Rule::UnnecessaryTypeUnion)
+              // Avoid duplicate checks if the parent is an `|`
+                && !matches!(
+                    checker.semantic.expr_parent(),
+                    Some(Expr::BinOp(ast::ExprBinOp { op: Operator::BitOr, ..}))
+                )
```

Comment:
Should we consider moving this check into a shared block like we do for the `Union[...]` case above?

Reply:
That would be nice, removing the duplicated code. Does the variable name `is_unchecked_union` fit here as well?